cmd/talosctl/cmd/talos/root.go

```diff
@@ -123,3 +127,92 @@ var Commands []*cobra.Command
 func addCommand(cmd *cobra.Command) {
 	Commands = append(Commands, cmd)
 }
+
+// completeResource represents tab complete options for `ls` and `ls *` commands.
+func completePathFromNode(inputPath string) []string {
+	pathToSearch := inputPath
+
+	// If the pathToSearch is empty, use root '/'
+	if pathToSearch == "" {
+		pathToSearch = "/"
+	}
+
+	var paths map[string]struct{}
+
+	// search up one level to find possible completions
+	if pathToSearch != "/" && !strings.HasSuffix(pathToSearch, "/") {
+		splits := strings.Split(pathToSearch, "/")
+		// we need a trailing slash to search for items in a directory
+		pathToSearch = strings.Join(splits[:len(splits)-1], "/") + "/"
+	}
+
+	paths = getPathFromNode(pathToSearch)
+
+	result := make([]string, 0, len(paths))
+
+	for k := range paths {
+		result = append(result, k)
+	}
+
+	return filterPaths(result, inputPath)
+}
+
+func filterPaths(input []string, filter string) []string {
+	result := make([]string, 0, len(input))
+
+	for _, v := range input {
+		if strings.HasPrefix(v, filter) {
+			result = append(result, v)
+		}
+	}
+
+	return result
+}
+
+//nolint:gocyclo
+func getPathFromNode(path string) map[string]struct{} {
+	paths := make(map[string]struct{})
+
+	if WithClient(func(ctx context.Context, c *client.Client) error {
+		stream, err := c.LS(ctx, &machineapi.ListRequest{
+			Root: path,
+		})
+		if err != nil {
+			return err
+		}
+
+		for {
+			resp, err := stream.Recv()
+			if err != nil {
+				if err == io.EOF || client.StatusCode(err) == codes.Canceled {
+					return nil
+				}
+
+				return fmt.Errorf("error streaming results: %s", err)
+			}
+
+			if resp.Metadata != nil && resp.Metadata.Error != "" {
+				continue
+			}
+
+			if resp.Error != "" {
+				continue
+			}
+
+			// // skip reference to the same directory
+			if resp.RelativeName == "." {
+				continue
+			}
+			// directories have a trailing slash
+			if resp.IsDir {
+				paths[path+resp.RelativeName+"/"] = struct{}{}
+			} else {
+				paths[path+resp.RelativeName] = struct{}{}
+			}
+		}
+	}) != nil {
+		return paths
+	}
```

Comment:
should we limit number of paths we look for? I don't know if it makes sense to add some hard limit on the paths we return to avoid iterating forever?

Reply:
fixed